Dropdown module

rehno-lindeque opened the pull request:
This is my "take" on the dropdown module in #11. It blew up into a much bigger project than I'd intended but once I pulled on the thread it just kept unspooling :man_shrugging:

To compare my changes to the original PR see: https://github.com/circuithub/elm-semantic-ui/compare/e2aa48f9603154ec44caaad0b9b603202fff44c4...circuithub:modules/dropdown

## Modules

* `SemanticUI.Modules.Dropdown`: Free-form dropdowns that are basically just toggles for their drawers. 
  * Typically used with drawers that contain only custom controls like checkboxes / radio buttons.
* `SemanticUI.Modules.Dropdown.Select`: Extends the dropdowns with "selectable" options.
  * Typically used to construct dropdown menus where the options correspond to actions rather than an active selection.
* `SemanticUI.Modules.Dropdown.Selection`: Similar to `Select`, but extends the dropdown even further with a current selection (`value`). Currently single selection only, but it appears reasonably straightforward to add multiple selection to this module in future.
  * Typically used as a form control, similar to `<input>` or `<select>`

## Remaining work

- [x] ~Cover dropdown icon with toggle area.~
    * Fixed by making the entire dropdown a toggle
- [x] ~Fix order of attributes~
    * Fixed by making button composition more clear and adding `uiClassList`
- [x] ~Buttons with just an icon currently render as a labeled icon button~
    * Fixed by overriding the element (html builder) passed into `Button.viewClass` and passing in labels separately in the way `Button` expects the label to be passed in.
- [x] ~Rename `SemanticUI.Modules.HtmlBuilder` to `Html.Builder`~
- [x] Nearly finished changing `Selection msg option` to `Selection msg option selection`. 
    * (Currently the selection type is `Maybe option`, which is a common situation but not sufficient for all cases. I prefer the api with `{ defaultLabel, selectionLabel }` where it works well, but it's too much of a compromise elsewhere.)


## Future work

* Currently the dropdown root is set as the toggle element by default. It's possible that in future we may want to prevent `on "click"` from bubbling up from the drawer up to the root. However to my surprise this click bubbling issue hasn't seemed to come up as a problem in any of the dropdowns we've got that has checkboxes etc. So I didn't want to fix what wasn't broken - but it's possible it could come up.
* Some of this ought to be split off into `Menu` and `Menu.Item`, but I'd prefer to leave that for future work since this is already a large PR.
* Also possibly there could be some kind of relationship with the `Label` element, but I haven't investigated
* Support `search` ("typeahead")
* It should probably be possible to specify more than one toggle event (e.g. `OnClick` as well as `OnFocus` probably makes sense, although I get the sense this may already be happening somehow in the underlying elm-dropdown module)
* Speaking of which, `OnFocus` is currently untested, it may require an `<input>` element to always be present or some other attribute - not sure. `OnHover` is also untested right now.
* Keyboard interaction needs work (tab, arrow keys, return).

## Related

Note the somewhat related changes in circuithub/elm-dropdown: https://github.com/circuithub/elm-dropdown/pull/1

domenkozar committed:
Add dropdown module

Most of code I've gotten from @ocharles, I've made some minor
corrections to documentation (typos, wrong examples).

domenkozar committed:
dropdown/select: make toggle icon overridable

rehno-lindeque committed:
Refactor: Eliminate the context type

rehno-lindeque committed:
An alternative take on the layout pattern in dropdown

rehno-lindeque committed:
Some subjective code formatting changes

rehno-lindeque committed:
HtmlRenderer to HtmlBuilder

ocharles commented on SemanticUI/Modules/Common.elm:
```diff
@@ -0,0 +1,9 @@
+module SemanticUI.Modules.Common exposing (HtmlBuilder)
```
I am not a fan of `Common` modules. It doesn't really have anything to do with `SemanticUI`, so I'd take it out of that namespace. How about just `Html.HtmlBuilder`?

rehno-lindeque replied:
Good idea, I like it.

rehno-lindeque replied:
Ah I missed the part about `Html.HtmlBuilder` namespace. Will do that

ocharles commented on SemanticUI/Modules/Dropdown.elm:
```diff
@@ -0,0 +1,438 @@
+dropdown :
+    Config msg
+    -> State
+    -> (DropdownBuilder msg -> html)
```
Could this be part of the `Config` for a `dropdown`?

rehno-lindeque replied:
It's a little awkward, but yes! Would need to be a slightly different type than for config given to `root` / `drawer` / `toggle` etc in that case.

I suppose I could go with

```elm
type alias Config config = {
  ...
}

type alias DropdownConfig = Config { layout : DropdownBuilder msg -> html }

init : { config | identifier : String, onToggle : State -> msg } -> Config config msg
init { identifier, onToggle } =
    { identifier = identifier
    , onToggle = onToggle
    , toggleEvent = OnClick
    , readOnly = False
    }

dropdownInit : { config | identifier : String, onToggle : State -> msg, layout : DropdownBuilder msg -> msg } -> Config config msg
dropdownInit { identifier, onToggle, layout } =
    { identifier = identifier
    , onToggle = onToggle
    , toggleEvent = OnClick
    , readOnly = False
    , layout = layout
    }


```

if that seems pleasant to use? It's not entirely clear to me... I'll circle back around to this after

ocharles replied:
It should be fine with just a single `Config` type, parameterised on `msg` and `html`.

rehno-lindeque replied:
I'm not sure if I follow? Currently the following top-level functions are exposed by the module:

```elm
root : Config msg -> State -> HtmlBuilder msg -> HtmlBuilder msg
drawer : Config msg -> State -> HtmlBuilder msg -> HtmlBuilder msg
toggle : Config msg -> State -> HtmlBuilder msg -> HtmlBuilder msg
```

where `Config msg` does not include the layout.

I'm not sure whether these ought to be exposed, but that's the situation as it stands. (This essentially lets you create toggles for the dropdown in completely unrelated parts of a document. It comes from elm-dropdown, not entirely sure whether it's necessary to preserve this for elm-semantic-ui :man_shrugging: )

rehno-lindeque replied:
This is probably no longer relevant, but sparked a bunch of follow up work.

ocharles reviewed (changes requested):
Two initial observations

rehno-lindeque committed:
WIP: customizable selectionable dropdowns

rehno-lindeque committed:
wip: Dropdown with a selection

rehno-lindeque committed:
WIP: tweaks / fixes

rehno-lindeque committed:
Move the layout functions into the config

rehno-lindeque committed:
Dropdown: Add drawer state directly to the config

rehno-lindeque committed:
Dropdown.Select: Add drawer state directly to the config

rehno-lindeque committed:
small naming tweak for the select component

rehno-lindeque committed:
Use toHtml style from SemanticUI.Element.List.Item to start simplifying types a little bit

rehno-lindeque commented on SemanticUI/Modules/Dropdown.elm:
```diff
@@ -0,0 +1,389 @@
+            toDropdown div
+                []
+                [ toToggle div [ class "text" ] [ text "File" ]
+                , toToggle i [ class "dropdown icon" ] []
```
I'm not sure if I understand why `toToggle` is needed on child elements of toDropdown? Is it that events don't always propagate up to `toDropdown`?

rehno-lindeque replied:
Changed this to `toToggle` on the root element. This needs some testing to confirm.

Possibly in future one could also forcibly stop on "click" event bubbling on the `drawer`, but for now I'm just doing the simplest possible thing until a reason comes up to change it.

rehno-lindeque committed:
don't expose the toggle function

rehno-lindeque committed:
It appears as though the drawer element will always be a <div> in SemanticUI

rehno-lindeque commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Dropdown/Select improvements](https://trello.com/c/cmZs0DTA/1206-dropdown-select-improvements)

rehno-lindeque commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Dropdown/Select improvements](https://trello.com/c/cmZs0DTA/1206-dropdown-select-improvements)

rehno-lindeque commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Dropdown/Select improvements](https://trello.com/c/cmZs0DTA/1206-dropdown-select-improvements)

rehno-lindeque committed:
Remove init and work with Dropdown & Select directly instead

rehno-lindeque committed:
Add a selection variant to Dropdown.Select

rehno-lindeque committed:
Add a labeled select dropdown and add selection labels to inline, selection templates

rehno-lindeque committed:
Add the 'default' class to the default label

rehno-lindeque committed:
small fixes & documentation polish

rehno-lindeque committed:
Add an optional dropdown icon to the select label

rehno-lindeque commented on SemanticUI/Modules/HtmlBuilder.elm:
```diff
@@ -0,0 +1,49 @@
+module SemanticUI.Modules.HtmlBuilder exposing (HtmlBuilder, appendChild, appendChildren, prependAttribute, prependAttributes, prependChild, prependChildren)
```
I'm not sure that this is the correct home for this module. Perhaps just `SemanticUI.HtmlBuilder`? (It's the module that used to be called `SemanticUI.Modules.Common`)

rehno-lindeque committed:
Add a simplified html builder  for Dropdown and Select

rehno-lindeque committed:
Distinguish between Labeled / Selection / Inline as mutually exclusive variations

rehno-lindeque committed:
Add fluid style to Select

rehno-lindeque committed:
Add the fluid style to Dropdown

rehno-lindeque committed:
Dropdown link item

rehno-lindeque committed:
select: fix fluid / selectattributes

rehno-lindeque committed:
Add a way to specify attributes even with simple html

rehno-lindeque committed:
Make "simple" html the default

rehno-lindeque committed:
Add the scrolling dropdown style

rehno-lindeque committed:
Add the compact style for selection

rehno-lindeque committed:
Make the compact style specific to the selection variation

rehno-lindeque committed:
Labeled applies to button dropdowns, this is a single-select

rehno-lindeque committed:
Add a button varation to dropdowns

rehno-lindeque commented on SemanticUI/Modules/Dropdown.elm:
```diff
@@ -206,15 +220,43 @@ dropdown config =
+                            (toRoot config element
+                                |> HtmlBuilder.prependAttributes attrs
+                            )
+                            buttonConfig
```
This is getting to the point where I'm getting pretty confused about the ordering of attributes.

I'd be happy to take suggestions on how to simplify. I think I'll need to review exactly how the ordering ends up

rehno-lindeque commented on SemanticUI/Modules/Dropdown.elm:
```diff
@@ -206,15 +220,43 @@ dropdown config =
+                            (toRoot config element
+                                |> HtmlBuilder.prependAttributes attrs
+                            )
+                            but
```
This is getting to the point where I'm getting pretty confused about the ordering of attributes.

I'd be happy to take suggestions on how to simplify. I think I'll need to review exactly how the ordering ends up

Update: indeed, ordering is not quite right, will need a bit of tweaking.

rehno-lindeque committed:
Add compact to button

rehno-lindeque committed:
A dropdown can be compact if it is a selection or a button

rehno-lindeque committed:
icon: fix 'left arrow' -> 'arrow left'

rehno-lindeque committed:
icon: add arrow up + arrow down

rehno-lindeque committed:
missing Select.button definition

rehno-lindeque committed:
Split dropdowns with stateful selection from selectable dropdowns

rehno-lindeque committed:
Split off Dropdown.Drawer and Dropdown.Toggle

domenkozar commented on SemanticUI/Modules/Dropdown.elm:
```diff
@@ -0,0 +1,434 @@
+            ]
+            []
+            :: children
+            ++ (if config.dropdownIcon then
```
It would be much better to allow specifying the icon. That's something we need, so a boolean is insufficient.

rehno-lindeque replied:
@domenkozar Do you have a button dropdown? 

If so, the way this seems to work is that it's technically a "labeled icon button" that also happens to be a dropdown. So you have:

`ui ...(dropdown classes)... dropdown ....(button classes)... button`

so for the icon:

`ui ...(dropdown classes)... dropdown labeled icon button`

From what I can see, SemanticUI appears to treat the `dropdown icon` in a special way - you can't seem to replace it directly without encountering layout bugs.

Here's how I'm doing it for a filter button at the moment:

```elm
            Selection.button
                { button =
                    Button.init
                        |> Button.icon (Just Icon.Filter)
                , identifier = "filter-by"
                , onToggle = SetFilterByDrawer
                , onSelect = SetFilterBy
                , drawerState = model.filterByDrawerState
                , currentSelection = Just model.selectedFilter
                , defaultLabel = text "Select filter"
                , selectionLabel = text << filterLabel
                }
                |> Selection.toHtml
                    { options = allFilters
                    , optionLabel = text << filterLabel
                    }
```

Does that support your use-case?

domenkozar replied:
I guess I'll play around once you have this ready and report back. Need to try to really say it works.

rehno-lindeque committed:
Make Selection module its own type

rehno-lindeque committed:
Move dropdown label, selection labels, formInput to their proper modules

rehno-lindeque committed:
missing exports

rehno-lindeque committed:
Simplify the dropdown toggle to fix dropdown icon being unclickable

rehno-lindeque committed:
Fix order of semantic ui classes

rehno-lindeque committed:
Fix labeled button generation

Don't include the drawer and other miscelaneous bits of html (hidden input etc) in the button label.

rehno-lindeque committed:
rename SemanticUI.Modules.HtmlBuilder to Html.Builder

HanStolpo commented on SemanticUI/Modules/Dropdown.elm:
```diff
@@ -0,0 +1,437 @@
+
+-}
+
+import Dropdown
```
I think this is not compiling at the moment am I correct ? Because there is no `Dropdown` module and no dependency exposing `Dropdown`, what is this `Dropdown` module and where is it coming from ?

rehno-lindeque replied:
It's still elm-dropdown as in the original PR, except that it's currently using https://github.com/circuithub/elm-dropdown/pull/1 mentioned in the related section above.

I didn't spend a ton of time looking at elm-dropdown, do you think there's any problems in it as it stands?

rehno-lindeque committed:
Fix missing custom config attributes

rehno-lindeque committed:
add pointer-events: none to allow search dropdown usage

rehno-lindeque committed:
parameterize by the selection (value) type

rehno-lindeque commented on SemanticUI/Modules/Dropdown/Selection.elm:
```diff
@@ -0,0 +1,399 @@
+single :
+    { config
+        | value : selection
+        , isSelected : option -> Bool
```
This `isSelected` config is probably the one config that I'm the most on the fence about, but I have no better ideas other than possibly:

* Renaming `isSelected` to `optionSelected` to indicate it's specifically on options
* Or changing it to `option -> selection -> Bool` so that one can use `isSelected = (==)` for single selection or `isSelected = List.elem` for multiple selection.
* Replacing with `toSelection : option -> selection` and then relying on `==` or `List.elem` internally.

None of them really stuck out to me so I just stayed with my original config. Happy to change this in future.

rehno-lindeque commented on SemanticUI/Modules/Dropdown/Selection.elm:
```diff
@@ -0,0 +1,399 @@
+        , onToggle : Drawer.State -> msg
+        , onSelect : option -> msg
+        , isSelected : option -> Bool
+        , formInput : Maybe { name : String, toValue : selection -> String }
```
`formInput` is the other thing I'm wondering whether we can do without. It seems like it's sort of mostly relevant for traditional form submission, but for modern elm it's all ajax anyway...

It's just that it's what semantic ui does to stay strictly compatible with other form elements I think...

Could use other's opinions if anyone has one

rehno-lindeque committed:
toToggle doesn't appear to be all that useful

That is SemanticUI already has a well defined structure for dropdowns.
The drawer state is also exposed, so there is always an escape hatch
available anyway.

rehno-lindeque committed:
A few documentation tweaks and caret rename

ocharles commented on SemanticUI/Modules/Dropdown/Select.elm:
```diff
@@ -0,0 +1,327 @@
+import SemanticUI.Modules.Dropdown.Toggle as Toggle
+
+
+{-| Select variations. For internal use only.
```
If it's for internal use, why is it exported?

rehno-lindeque replied:
Fair point... I think this comment should probably read similar to the comment above `Config` instead of this.

The export is an escape hatch in case people want to just deal directly with the config rather than the carefully thought out constructors defined here (`inline` / `button` / etc).

ocharles commented on SemanticUI/Modules/Dropdown/Select.elm:
```diff
@@ -0,0 +1,327 @@
+
+
+variationClassList : Variation msg -> List ( String, Bool )
+variationClassList variation =
```
How about `variationAttributes : Variation msg -> List (Attribute msg)`? I suggest this because it tidies up the `(  "selection", True )` element, into just `class "selection"`.

rehno-lindeque replied:
Variations actually have attributes and classes which I started dealing with separately starting from this commit: https://github.com/circuithub/elm-semantic-ui/commit/d9a54d54c35ce581ee2d641aee4b5771cc76eea0.

The main problem is that you have to be rather careful with the order of classes and if memory serves me it's actually sort of the opposite (in semantic ui) of what you want for attributes in general.

In theory I could make this `List [ String ]`, but `classList` feels to me a bit more idiomatic to elm.

ocharles commented on SemanticUI/Modules/Dropdown/Select.elm:
```diff
@@ -0,0 +1,327 @@
+
+{-| Everything needed to build the `Html msg` representation of a particular select dropdown.
+
+  - toDropdown - converts a `<div>` or an `<a>` element into a SemanticUI dropdown
```
`toDropdown` works with anything, not just `div` or `a`:

```suggestion
  - toDropdown - converts a HTML element into a SemanticUI dropdown. Often used with `div` or `a`.
```

rehno-lindeque replied:
I agree with the change, but just to mention, this is the list of elements where I know for sure that the `dropdown` class will have consistent styling (e.g. with corrected margins on the caret and misc stuff like that). Not a big deal, but just to explain the reasoning behind the comment.

rehno-lindeque replied:
The other reason for the comment is sort of as preparation for properly supporting other dropdown root elements. (I think `<input>` (for search selection) and possibly `<select>` can also be used, but I need to check again.)

rehno-lindeque replied:
Oh I nearly forgot, but `selection` is actually the tricky one. It connects the dropdown toggle element to the drawer in a way that will only layout as intended if you've picked an appropriate root element. So I wanted to point people to `<div>` by default, and `<a>` if you're building nested menus...

ocharles commented on SemanticUI/Modules/Dropdown/Select.elm:
```diff
@@ -0,0 +1,327 @@
+{-| Everything needed to build the `Html msg` representation of a particular select dropdown.
+
+  - toDropdown - converts a `<div>` or an `<a>` element into a SemanticUI dropdown
+  - toOption - takes the option value and converts a `<div>` or an `<a>` element into a selectable option item
```
```suggestion
  - toOption - takes a dropdown option value and renders it as a HTML element that is a selectable option item
```

rehno-lindeque replied:
~This is a case where I think only `<div>` and `<a>` really makes sense for the `item` class. Other elements can be used, but they won't layout consistently afaik.~

Well... on second thought, this is probably not true, it looks like many elements do layout fine. It's just that the examples all use `div` or `a`, depending on the type of menu being constructed. It seems unlikely that you'll need a different element.

ocharles commented on SemanticUI/Modules/Dropdown/Select.elm:
```diff
@@ -0,0 +1,327 @@
+
+  - toDropdown - converts a `<div>` or an `<a>` element into a SemanticUI dropdown
+  - toOption - takes the option value and converts a `<div>` or an `<a>` element into a selectable option item
+  - drawer - the drawer element that can be toggled open or closed
```
It wasn't immediately obvious to me what a drawer was. How about

```suggestion
  - drawer - the drawer element contains all of the selects options
```

?

rehno-lindeque replied:
I think in the context of a `Dropdown.Select` this is appropriate yes :+1:.

Side note: I noticed that material ui seems to be good source for some of these types of definitions btw. https://material-ui.com/api/drawer/

rehno-lindeque replied:
(Now that I look more closely I see that in material ui this refers to a navigation (side bar) drawer. I think the same terminology does apply to dropdowns but I can't tell you where I'm getting that from...)

ocharles commented on SemanticUI/Modules/Dropdown/Select.elm:
```diff
@@ -0,0 +1,327 @@
+{-| A dropdown component with selectable options.
+
+This is a vanilla select component that can be extended in many different ways (including multiple selection).
+It does not highlight the current selection automatically, see `SemanticUI.Modules.Dropdown.Selection` for this.
```
Does "automatically" mean "when the drawer is open"?

rehno-lindeque replied:
It means that it's possible to apply the `active` class to the selected options by applying some custom layout but it doesn't do so automatically. You need `Selection` for that. (Seems to be semantic ui's standard way of highlighting the current selection(s) in the drawer menu.) 

I can improve the wording here, it's a little beside the point.

rehno-lindeque replied:
Probably better explained at the top of the file https://github.com/circuithub/elm-semantic-ui/pull/21/files#diff-203e3a171f8134454e49702893608197R21-R36, I'll remove this.

ocharles reviewed (commented):
A few comments as I review the `Dropdown.Select` module. My main hesitation is still the separation between the dropdown and `Html`. I note that `select : { config | label : Maybe (Html msg }`, but then `toHtml` also requires `Html` producing functions. It also seems strange that `toHtml` needs the list of options, and not the definition of the control.

To me, a more intuitive API would be:

```elm
select : { config | options : List option, ... } -> Select option msg
toDiv : { optionToHtml : option -> Html msg } -> Select option msg
```

That is, `select` (and `button`) mentions nothing about `Html`, this only happens in the `toDiv` function.

Also, I would rename `toHtml` -> `toDiv`, and `toCustomHtml` -> `toHtml`.

rehno-lindeque commented:
Just to quickly mention the evolution of the API, I started off with:

```elm
toHtml : Builder -> Html msg
toSimpleHtml : { options : List (Html msg), optionLabel : option -> Html msg }
```

I decided to swap the two to optimize for the common use case (where you don't need to customize the layout of the dropdown drawer menu).


> A few comments as I review the `Dropdown.Select` module. My main hesitation is still the separation between the dropdown and `Html`. I note that `select : { config | label : Maybe (Html msg }`, but then `toHtml` also requires `Html` producing functions. It also seems strange that `toHtml` needs the list of options, and not the definition of the control.

I think we're going to find that we do need `Html` types in the config, but that it's often a placeholder for some subcomponent that needs more work. Additionally the `Builder` is a bit of a red herring and should be pushed down into sub components like `Menu` instead. (It was a little confusing starting from the original PR with all the flexibility it provides.)

To be totally honest, I can't say I'm settled on the drawer menu either. I vaguely mention something like this in "Future Work" but I think the nicest api might actually be something along the lines of an option `Menu`:

```elm

-- Drawer as a menu
let drawerMenu = Menu.simpleMenu { ... }
Dropdown.dropdown { ...., drawer = Drawer.asMenu menu }
    |> Dropdown.toHtml

-- Drawer as any arbitrary thing (not necessarily a menu).

let drawerContents = div [..] [..]
Dropdown.dropdown { ...., drawer = Drawer.asHtml drawerContents }
    |> Dropdown.toHtml
```

> To me, a more intuitive API would be:
> 
> ```elm
> select : { config | options : List option, ... } -> Select option msg
> toDiv : { optionToHtml : option -> Html msg } -> Select option msg
> ```
>
> That is, `select` (and `button`) mentions nothing about `Html`, this only happens in the `toDiv` function.
> 
> Also, I would rename `toHtml` -> `toDiv`, and `toCustomHtml` -> `toHtml`.

The problem is mainly that `options` only really makes sense if you're using the simplest layout. A custom layout is probably not going to want a list of options directly since it's going to be interspersing things like dividers and headings.

E.g.

<img width="402" alt="ui-complex-menu" src="https://user-images.githubusercontent.com/337811/66771287-d9961c80-ee87-11e9-9f18-d22893d931d1.png">

---

I'm just unfortunately so far behind on a lot of work that I'm not really up to spending a week looking at something like `Menu` in detail right now, but I think it does need doing.

Seeing as the original PR was close to merging, would you be ok with this one as well? Even though the drawer menu is not exactly ideal, this PR solves a bunch of problems with `Dropdown.Select` in particular. I think that what this PR offers is that it breaks down the variations in a way that can be refactored much more uniformly in the future, in contrast to the previous iteration of this code. 

It's also now possible to extend it with things like multiple selection, or with `RemoteData` as the `selection` type and things of that sort without too much hassle.

rehno-lindeque commented on SemanticUI/Modules/Dropdown/Select.elm:
```diff
@@ -0,0 +1,327 @@
+It does not highlight the current selection automatically, see `SemanticUI.Modules.Dropdown.Selection` for this.
+
+-}
+select :
```
Looking at this yet again, the `select` naming is still confusing. I'm going to rename it.

rehno-lindeque replied:
Renamed to `Select.ordinary`

rehno-lindeque committed:
some documentation fixes for the PR

rehno-lindeque committed:
Rename Select.select -> Select.ordinary

rehno-lindeque committed:
Merge branch 'master' of github.com:circuithub/elm-semantic-ui into modules/dropdown

rehno-lindeque committed:
make elm-dropdown dependency explicit

rehno-lindeque committed:
Move Html.Builder module into src/

rehno-lindeque committed:
Fix rename issue that sneaked in

rehno-lindeque commented:
Oops I'd forgotten to actually push to master, it's merged now
